src/lib/middleware/rateLimiting.ts

```diff
@@ -15,6 +19,37 @@ const {
   ENABLE_RATE_LIMITING,
 } = config
 
+export const BURST_LIMIT_MESSAGE = "Too Many Requests (3.1)"
+export const RATE_LIMIT_MESSAGE = "Too Many Requests (3.2)"
+
+/**
+ * If ddos traffic gets into force this is our first line of defense.
+ * Every cycle counts. It processes the burst limiter first which is
+ * the most performant and most likely to be tripped by an attack. It
+ * goes through the rateLimiter afterwards for a marginally slower, but
+ * distributed rate limit.
+ */
+export const _rateLimiterMiddleware = <
+  B extends RateLimiterAbstract,
+  R extends RateLimiterAbstract
+>(
+  burstLimiter: B,
+  rateLimiter: R
+) => (req, res, next) => {
+  const ip = requestIp.getClientIp(req)
+
+  // Ensure promise returns for test purposes
+  // Not needed for actual middleware
+  return burstLimiter.consume(ip).then(
+    () =>
+      rateLimiter
+        .consume(ip)
+        .then(() => next())
```

Comment:
Very tiny, but perhaps it counts when we’re trying to work with many requests; we can get rid of this extra per-request closure by just passing `next` to `then`.